kernel/src/net/socket/util/message_header.rs

```diff
@@ -1,32 +1,190 @@
 // SPDX-License-Identifier: MPL-2.0
 
+use align_ext::AlignExt;
+
 use super::SocketAddr;
-use crate::prelude::*;
+use crate::{net::socket::unix::UnixControlMessage, prelude::*, util::net::CSocketOptionLevel};
 
 /// Message header used for sendmsg/recvmsg.
 #[derive(Debug)]
 pub struct MessageHeader {
     pub(in crate::net) addr: Option<SocketAddr>,
-    pub(in crate::net) control_message: Option<ControlMessage>,
+    pub(in crate::net) control_messages: Vec<ControlMessage>,
 }
 
 impl MessageHeader {
     /// Creates a new `MessageHeader`.
-    pub const fn new(addr: Option<SocketAddr>, control_message: Option<ControlMessage>) -> Self {
+    pub const fn new(addr: Option<SocketAddr>, control_messages: Vec<ControlMessage>) -> Self {
         Self {
             addr,
-            control_message,
+            control_messages,
         }
     }
 
     /// Returns the socket address.
     pub fn addr(&self) -> Option<&SocketAddr> {
         self.addr.as_ref()
     }
+
+    /// Returns the control messages.
+    pub fn control_messages(&self) -> &Vec<ControlMessage> {
+        &self.control_messages
+    }
 }
 
-/// Control message carried by MessageHeader.
-///
-/// TODO: Implement the struct. The struct is empty now.
+/// Control messages in [`MessageHeader`].
 #[derive(Debug)]
-pub struct ControlMessage;
+pub enum ControlMessage {
+    Unix(UnixControlMessage),
+}
+
+impl ControlMessage {
+    // FIXME: This method may exhaust kernel memory and cause a panic if the program is malicious
+    // and attempts to send too many control messages. To prevent this, we can either limit the
+    // number of control messages or aggregate control messages of the same type to avoid the
+    // vector.
```

Comment:
In fact, when I read the code again, I think we should just read to a structure similar to `AuxiliaryData` instead of storing it in the vector, so we can resolve this FIXME. Let me try to see if it works.

Reply:
Well. After conducting some experiments, I'm not sure if this works well.

First, it costs more to pass around a structure like `AuxiliaryData` than it does to pass around a structure like `Vec`. We may not worry about that much now, but it will be a problem if we have more IP/TCP/UDP control messages in the future. One alternative is to use something like `Option<Box<AuxiliaryData>>`, but that seems complicated.

Another issue is that it interacts with credential processing in https://github.com/asterinas/asterinas/pull/2268. I have not decided what's the best way to deal with it now. So maybe we should investigate this in the future.

Reply:
I’m fine with simply putting a FIXME here, since memory exhaustion is a rare case.